### PR TITLE
feat: [😭] Firearm `AmmoModules` Api

### PR DESCRIPTION
**Pull request by IRacle1**
## Description
**Describe the changes** 

*inspired by* #255 

I was thinking about more flexible logic for setting ammo. My first idea was to make better abstractions of `Barrel` and `Primary` magazines(that what i did in that pr firstly). I liked that idea, and that gave me a possibility to add more specific api for specific magazines (`Rotate` for revolvers, or bolt locking for automatic firearms as example). But there is 2 questions about ammo setters and for modules in general:
 - there was a discussion in 255 pr, [taptap](https://github.com/ExMod-Team/EXILED/pull/255#discussion_r1863846886) , for giving support to more than one primary/barrel magazines at ones. Ive thought about it later, and Ive decided in this case to move away from several magazines abstraction in favour of easier system of maximum 2 magazines (barrel and primary). But if adding custom modules turns to be real and possible, were gonna have to look at this again.
 - main question is an about ammo setter at all, do we generally need such complicated `Ammo`/`MaxAmmo` setters? When i finished with main part of api, and started thinking about total `Ammo` setter. And I really dont think an absolute/total ammo setter like before 14.0 is needed. This setter has an overly complex background and logic. Even if we position Exiled as a highlevel api, I still dont think its necessary to make a single setter for barrel and magazine ammos. But anyway that my honest shizo opinion says. Really wait for feedback😘

*oh, and i added RotatingRevolver event😎*

**What is the current behavior?** (You can also link to an open issue here)
😥

**What is the new behavior?** (if this is a feature change)
😎

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
😓

**Other information**:
🥹
<br />

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentations
<br />

## Submission checklist

- [x] I have checked the project can be compiled
- [x] I have tested my changes and it worked as expected

### Patches (if there are any changes related to Harmony patches)
- [x] I have checked no IL patching errors in the console

## Main Axioms that i used for api logic, or tldr😎
- `Barrel` and `Normal` magazines should be clearly segregated for the convenience of people
- In those abstraction `Ammo` setters should be available, despite the lack of this abstraction in the basegame.
- As i think now, the best way will to create `BarrelAmmo` `MagazineAmmo` full properties, and make `TotalAmmo` only with getter

### CODE REVIEW

**Comment by IRacle1** [comment]
All ammoModules🥹
```yaml
firearm: GunCOM15
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunCOM18
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunE11SR
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunCrossvec
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunFSP9
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunLogicer
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunRevolver
 - primary: CylinderAmmoModule
firearm: GunAK
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunShotgun
 - barrel: PumpActionModule
 - primary: MagazineModule
firearm: GunCom45
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: ParticleDisruptor
 - primary: MagazineModule
firearm: GunFRMG0
 - barrel: AutomaticActionModule
 - primary: MagazineModule
firearm: GunA7
 - barrel: AutomaticActionModule
 - primary: MagazineModule
```

**Comment by IRacle1** [comment]
Anyway until we come to an agreement on that Ill leave it to the draft

**Comment by Misaka-ZeroTwo** [comment]
I like the idea of separating the barrel and the magazine. 

It will be way too complex to handle them all in one place. We would deal with more edge cases and a much higher chance that something just doesn't work.

**Comment by VALERA771** [comment]
Lgtm. For some people may be too complex to use but game the game is also is not too easy to understand. So if people don't need to use a lot of classes to get a ammo and and if we don't have a 70 line ammo setter I agree

**Comment by VALERA771** [comment]
Also waiting for these guys opinions

File: EXILED/Exiled.API/Features/Items/Firearm.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:193]: Inlined this

File: EXILED/Exiled.API/Features/Items/Firearm.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:206]: Inlined these

File: EXILED/Exiled.API/Features/Items/Firearm.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:215]: My brain didn't get what these remarks mean and why 2 is needed

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:50]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:68]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:89]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:99]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/PumpBarrelMagazine.cs:111]: Inlined these

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]: Can be inlined

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:107]: Inlined these

**Review by louis1706 — CHANGES_REQUESTED** [review]
I am lazzy to finish just inlined everything than can be inlined

File: EXILED/Exiled.API/Features/Items/Firearm.cs
Comment (IRacle1) [reply EXILED/Exiled.API/Features/Items/Firearm.cs:215]: Oh, i forgot to set it for 1 by default, i used `2` for testing😭

File: EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs
Comment (IRacle1) [reply EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]: How🤨

Comment (louis1706) [reply EXILED/Exiled.API/Features/Items/FirearmModules/Magazine.cs:96]:
```suggestion
            ammo = useBorders ? Mathf.Clamp(Ammo + delta, 0, MaxAmmo) : Ammo + delta;
```

**Comment by VALERA771** [comment]
Is this PR is ready for reviews and tests? I think it's one the best one from other PRs which are fixing firearm properties

**Comment by VALERA771** [comment]
Some things (a lot) can be inlined. But generally looks good if you've tested magazine API

File: EXILED/Exiled.API/Features/Items/Firearm.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/Firearm.cs:168]:
```suggestion
        public int TotalAmmo => Base.GetTotalStoredAmmo();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/AutomaticBarrelMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Barrel/AutomaticBarrelMagazine.cs:66]:
```suggestion
            set => AutomaticBarrel.ChamberSize = Mathf.Clamp(value, 0, 16);
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:58]:
```suggestion
            get => Magazine.AmmoType.GetAmmoType();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:63]:
```suggestion
            set => MagazineModule._ammoType = value.GetItemType();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:74]:
```suggestion
            get => MagazineModule.MagazineInserted;
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:93]:
```suggestion
        public void RemoveMagazine() => MagazineModule.ServerRemoveMagazine();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:101]:
```suggestion
        public void InsertMagazine() => MagazineModule.ServerInsertEmptyMagazine();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/NormalMagazine.cs:107]:
```suggestion
        public override void Resync() => MagazineModule.ServerResyncData();
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs:44]:
```suggestion
        public override int MaxAmmo => Magazine.AmmoMax;
```

File: EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs
Comment (louis1706) [inline EXILED/Exiled.API/Features/Items/FirearmModules/Primary/PrimaryMagazine.cs:57]:
```suggestion
            get => Magazine.AmmoStored;
```

File: EXILED/Exiled.Events/EventArgs/Player/RotatingRevolverEventArgs.cs
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Player/RotatingRevolverEventArgs.cs:54]:
```suggestion
        public bool HasEffect => Firearm.PrimaryMagazine.MaxAmmo % Rotations == 0;
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:62]: Is it possible to comment this part of code ? to explain more what it's does

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:79]:
```suggestion
            newInstructions.InsertRange(0, GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:92]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:104]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:126]:
```suggestion
            newInstructions.InsertRange(0, GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:139]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:151]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:178]:
```suggestion
            newInstructions.InsertRange(0, AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:191]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:203]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:230]:
```suggestion
            newInstructions.InsertRange(0, AmmoDrainAutomatic.GetInstructions(firearm, ammoDrain, cnt));
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:244]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Mul),
                });
```

File: EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/AmmoDrain.cs:256]:
```suggestion
            newInstructions.InsertRange(index, new CodeInstruction[]
                {
                    new(OpCodes.Ldloc_S, ammoDrain.LocalIndex),
                    new(OpCodes.Div),
                });
```

**Review by louis1706 — APPROVED** [review]
It's all fine to merge when this will be completed (it's can be done after it's merge if needed)

the only important it's for AmmoDrain.cs file about comment on the GetInstructions method

**Review by BoltonDev — APPROVED** [review]

**Comment by BonesMgee** [comment]
Epic 🧀

**Pull request merged**
